### PR TITLE
add a RW lock for duties

### PR DESCRIPTION
**Pull request by qu0b**
**What type of PR is this?**
> Bug fix

**What does this PR do? Why is it needed?**

This PR adds a RW lock for modifying and accessing the `duties` struct in the `validator` struct. Together with the [antithesis](https://antithesis.com/) team we discovered a race condition where two go routines are trying to access the same memory location related to the `v.duties` struct. The race condition can lead to a panic exiting the validator process.

**Which issues(s) does this PR fix?**

The data race before the panic:
```
[ 1320810] [  1667.682068] [service_prysm-geth-0--prysm-vc]             [I] WARNING: DATA RACE
[ 1320810] [  1667.682078] [service_prysm-geth-0--prysm-vc]             [I] Read at 0x00c0003778f8 by goroutine 1235:
[ 1320810] [  1667.682088] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.(*validator).duty()
[ 1320810] [  1667.682088] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/attest.go:271 +0xde
[ 1320810] [  1667.682094] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.(*validator).SubmitSignedContributionAndProof()
[ 1320810] [  1667.682094] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/sync_committee.go:126 +0x2c4
[ 1320810] [  1667.682100] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.performRoles.func1()
[ 1320810] [  1667.682100] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/runner.go:345 +0x584
[ 1320810] [  1667.682105] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.performRoles.func3()
[ 1320810] [  1667.682105] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/runner.go:353 +0xa1
[ 1320810] [  1667.683551] [service_prysm-geth-0--prysm-vc]             [I] Previous write at 0x00c0003778f8 by goroutine 249:
[ 1320810] [  1667.683556] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.(*validator).UpdateDuties()
[ 1320810] [  1667.683556] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/validator.go:845 +0xac4
[ 1320810] [  1667.683562] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.run()
[ 1320810] [  1667.683562] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/runner.go:121 +0x1418
[ 1320810] [  1667.683567] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.(*ValidatorService).Start.func1()
[ 1320810] [  1667.683567] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/service.go:250 +0x64
[ 1320810] [  1667.683578] [service_prysm-geth-0--prysm-vc]             [I] Goroutine 1235 (running) created at:
[ 1320810] [  1667.683583] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.performRoles()
[ 1320810] [  1667.683583] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/runner.go:327 +0x37e
[ 1320810] [  1667.683589] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.run()
[ 1320810] [  1667.683589] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/runner.go:174 +0x1a31
[ 1320810] [  1667.683594] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.(*ValidatorService).Start.func1()
[ 1320810] [  1667.683594] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/service.go:250 +0x64
[ 1320810] [  1667.683605] [service_prysm-geth-0--prysm-vc]             [I] Goroutine 249 (running) created at:
[ 1320810] [  1667.683610] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.(*ValidatorService).Start()
[ 1320810] [  1667.683610] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/service.go:250 +0x14ee
[ 1320810] [  1667.683616] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/runtime.(*ServiceRegistry).StartAll.func1()
[ 1320810] [  1667.683616] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/runtime/service_registry.go:42 +0x48
[ 1320810] [  1667.683621] [service_prysm-geth-0--prysm-vc]             [I] ==================
```

The panic (The log lines are off because of the instrumentation added): 
```
[ 1320810] [  1667.724495] [service_prysm-geth-0--prysm-vc]             [I] panic: runtime error: invalid memory address or nil pointer dereference
[ 1320810] [  1667.724538] [service_prysm-geth-0--prysm-vc]             [I] [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x232ab10]
[ 1320810] [  1667.724567] [service_prysm-geth-0--prysm-vc]             [I] goroutine 1234 [running]:
[ 1320810] [  1667.725164] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.(*validator).duty(0xc000377880, {0xb2, 0x12, 0x4e, 0x1b, 0x55, 0x43, 0x13, 0x17, 0xe2, ...})
[ 1320810] [  1667.725197] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/attest.go:271 +0xf0
[ 1320810] [  1667.725821] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.(*validator).SubmitSignedContributionAndProof(0xc000377880, {0x3006c00, 0xc000c1cf00}, 0x5f, {0xb2, 0x12, 0x4e, 0x1b, 0x55, 0x43, ...})
[ 1320810] [  1667.725855] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/sync_committee.go:126 +0x2c5
[ 1320810] [  1667.726443] [service_prysm-geth-0--prysm-vc]             [I] github.com/prysmaticlabs/prysm/v4/validator/client.performRoles.func1(0x5, {0xb2, 0x12, 0x4e, 0x1b, 0x55, 0x43, 0x13, 0x17, 0xe2, ...})
[ 1320810] [  1667.726476] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/runner.go:345 +0x585
[ 1320810] [  1667.726491] [service_prysm-geth-0--prysm-vc]             [I] created by github.com/prysmaticlabs/prysm/v4/validator/client.performRoles
[ 1320810] [  1667.726525] [service_prysm-geth-0--prysm-vc]             [I] /git/prysm_instrumented/customer/validator/client/runner.go:327 +0x37f
```

Line 845 in validator.go:
```go
842 resp, err := v.validatorClient.GetDuties(ctx, req)
843 if err != nil {
844     __antithesis_instrumentation__.Notify(72630)
845     v.duties = nil
        log.Error(err)
        return err
    } else {
        __antithesis_instrumentation__.Notify(72631)
    }
```

attest.go line 271 of the instrumented code
```go
271 for _, duty := range v.duties.Duties {
272       __antithesis_instrumentation__.Notify(70002)
        if bytes.Equal(pubKey[:], duty.PublicKey) {
            __antithesis_instrumentation__.Notify(70003)
            return duty, nil
        } else {
            __antithesis_instrumentation__.Notify(70004)
        }
    }
 ```

### CODE REVIEW

File: validator/client/validator.go [hunk 1]
```diff
@@ -745,10 +752,12 @@ func (v *validator) RolesAt(ctx context.Context, slot primitives.Slot) (map[[fie
 		// the validator checks whether it's in the sync committee of following epoch.
 		inSyncCommittee := false
 		if slots.IsEpochEnd(slot) {
+			v.dutiesLock.RLock()
 			if v.duties.NextEpochDuties[validator].IsSyncCommittee {
 				roles = append(roles, iface.RoleSyncCommittee)
 				inSyncCommittee = true
 			}
+			v.dutiesLock.RUnlock()
```
Comment (prestonvanloon) [inline validator/client/validator.go:760]: You already have a RLock at the top of this method.

This is probably what is causing the static analysis check. 

```
validator/client/validator.go:755:4: found recursive read lock call (recursivelock)
```

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Fails to build due to static analysis check

File: validator/client/validator.go (hunk 1 above)
Comment (qu0b) [reply validator/client/validator.go]: @prestonvanloon thanks for the feedback. I've updated the locks and hope that there are no more recursive calls.

**Review by prestonvanloon — APPROVED** [review]
Looks great. Thanks so much!

**Pull request merged**